akka-actor/src/main/scala/akka/util/WildcardTreeIndex.scala

```diff
@@ -0,0 +1,54 @@
+/**
+ * Copyright (C) 2009-2016 Lightbend Inc. <http://www.lightbend.com>
+ */
+
+package akka.util
+
+import scala.annotation.tailrec
+import scala.collection.immutable.HashMap
+
+private[akka] final case class WildcardIndex[T](wildcardTree: WildcardTree[T] = WildcardTree[T](), doubleWildcardTree: WildcardTree[T] = WildcardTree[T]()) {
+
+  def insert(elems: Array[String], d: T): WildcardIndex[T] = elems.lastOption match {
+    case Some("**") ⇒ copy(doubleWildcardTree = doubleWildcardTree.insert(elems.iterator, d))
+    case Some(_)    ⇒ copy(wildcardTree = wildcardTree.insert(elems.iterator, d))
+    case _          ⇒ this
+  }
+
+  def find(elems: Iterable[String]): Option[T] =
+    wildcardTree.findWithSingleWildcard(elems.iterator)
+      .orElse(doubleWildcardTree.findWithTerminalDoubleWildcard(elems.iterator))
+      .flatMap(_.data)
+}
+
+private[akka] final case class WildcardTree[T](data: Option[T] = None, children: Map[String, WildcardTree[T]] = HashMap[String, WildcardTree[T]]()) {
+
+  def insert(elems: Iterator[String], d: T): WildcardTree[T] =
+    if (!elems.hasNext) {
+      copy(data = Some(d))
+    } else {
+      val e = elems.next()
+      copy(children = children.updated(e, children.getOrElse(e, WildcardTree[T]()).insert(elems, d)))
+    }
+
+  @tailrec def findWithSingleWildcard(elems: Iterator[String]): Option[WildcardTree[T]] =
```

Comment:
The old implementation used an empty `WildcardTree` to indicate no-match. I think we can use the same here to avoid the `Some` allocations.